File: Cpp/gmock_version/tyrepressure/Alarm.cpp

#include "Alarm.hpp"
#include "Sensor.hpp"

Alarm::Alarm(Sensor *sensor) : m_sensor(sensor),
                               m_lowPressureThreshold(17),
                               m_highPressureThreshold(21),
                               m_alarmOn(false)
{}

void Alarm::check()
{
    double psiPressureValue = m_sensor->popNextPressurePsiValue();

    if (psiPressureValue < m_lowPressureThreshold || m_highPressureThreshold < psiPressureValue)
    {
        m_alarmOn = true;
    }
}

bool Alarm::isAlarmOn()
{
    return m_alarmOn;
}

File: CSharp/TirePressureMonitoringSystem/Alarm.cs
namespace TDDMicroExercises.TirePressureMonitoringSystem
{
    public class Alarm
    {
        readonly ISensor _sensor;
        private Threshold threshold;

        public Alarm(ISensor sensor, Threshold threshold)
        {
            this._sensor = sensor;
            this.threshold = threshold;
        }

        public void Check()
        {
            double psiPressureValue = _sensor.PopNextPressurePsiValue();

            if (psiPressureValue < threshold.Min || threshold.Max < psiPressureValue)
            {
                this.AlarmOn = true;
            }
        }

        public bool AlarmOn
        {
            get; private set;
        }

    }

    public class Threshold
    {

        public double Min
        {
            get;
            private set;
        }

        public double Max
        {
            get;
            private set;
        }

        public Threshold(double minThreshold, double maxThreshold)
        {
            this.Min = minThreshold;
            this.Max = maxThreshold;
        }
    }
}
